DEPR: Add deprecation warning for factorize() order keyword

EricChea opened the pull request:
#### What's this PR do?
In response to #19727.
Adds functionality to the ``deprecate_kwargs`` utility.  Currently, the utility doesn't allow the deprecation of an old keyword without having a new one in place.  The change proposed here allows the deprecation of an *old* keyword without specifying a *new* keyword by allowing users to pass ``new_arg_name=None``  to ``deprecate_kwargs``. 

#### PR checklist
- [X] closes #19727
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

EricChea committed:
[Utils][Warnings] Add deprecation warning for factorize() keyword, order. Add deprecation warning for entire keywords to deprecate_kwaargs util.

EricChea commented on pandas/tests/util/test_util.py:
```diff
@@ -34,9 +34,14 @@ def _f2(new=False):
+        def _f4(old=True, unchanged=True):
+            return old
+
         self.f1 = _f1
```
In a nutshell, to test that warnings are only raised if ``_f4()`` is called with the ``old`` keyword (*keyword to deprecate*).

I gave ``_f4()`` two keywords: 
- **old** to represent the keyword that's not removed yet.
- **unchanged** is meant to represent a keyword that shouldn't be impacted by the decorator.

With f4 I can test that:
- A *FutureWarning* will get raised if a user calls f4 with the old (*deprecated keyword*). For example ``self.f4(old=9)`` will raise an alert that the ``old`` keyword is no longer supported.
- No warning is raised if f4 is called without the old keyword.  For example, ``self.f4(unchanged=x)`` will not raise an alert.

Hope that makes sense.

jreback commented on pandas/util/_decorators.py:
```diff
@@ -65,8 +65,9 @@ def deprecate_kwarg(old_arg_name, new_arg_name, mapping=None, stacklevel=2):
-    new_arg_name : str
-        Name of preferred argument in function
+    new_arg_name : str | None
+        Name of preferred argument in function. Use none to raise warning that
```
none -> None

EricChea replied:
Thanks for catching that.  Please refer to 699805198f7e31e3f6787b612ec507dd71c056a6

jreback commented on pandas/util/_decorators.py:
```diff
@@ -107,6 +108,16 @@ def _deprecate_kwarg(func):
+
+            if new_arg_name is None and old_arg_value is not None:
+                msg = (
+                    "the '{old_name}' keyword is no longer supported"
```
the {old_name} keyword is deprecated and will be removed in a future version

 is our typical phrasing here.

EricChea replied:
Thank you.  Please refer to 699805198f7e31e3f6787b612ec507dd71c056a6

jreback commented on pandas/core/algorithms.py:
```diff
@@ -436,6 +437,7 @@ def isin(comps, values):
     return f(comps, values)
 
 
+@deprecate_kwarg(old_arg_name='order', new_arg_name=None)
```
ca you add a test which catches this, AND see if we have any existing cases of actually passing this arg

EricChea replied:
Sure thing.  I added two additional tests.  Please refer to 699805198f7e31e3f6787b612ec507dd71c056a6.

I did not find any cases in the Repo where ``factorize()`` is called with an ``order`` parameter.

jreback reviewed (changes requested):
need to add a deprecation notice in 0.23.0 whatsnew

EricChea committed:
PR review for deprecating keyword, order, from factorize().

  - Change none to None in docs.
  - Change warning message to be consistent with other phrasing.
  - Add tests to catch FutureWarning when order keyword is passed to factorize.

pep8speaks commented:
Hello @EricChea! Thanks for updating the PR.







 - In the file [`pandas/util/_decorators.py`](https://github.com/pandas-dev/pandas/blob/699805198f7e31e3f6787b612ec507dd71c056a6/pandas/util/_decorators.py), following are the PEP8 issues :

> [Line 114:80](https://github.com/pandas-dev/pandas/blob/699805198f7e31e3f6787b612ec507dd71c056a6/pandas/util/_decorators.py#L114): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 79 characters)

jreback commented on pandas/tests/test_algos.py:
```diff
@@ -248,6 +248,12 @@ def test_uint64_factorize(self):
         tm.assert_numpy_array_equal(uniques, exp_uniques)
 
+    def test_deprecate_order(self):
+        data = np.array([2**63, 1, 2**63], dtype=np.uint64)
```
can you add the gh issue number here as a comment

EricChea replied:
Good point.  I added an additional comment so we can remember to deprecate the test once the ``order`` keyword is finally removed.

jreback commented on pandas/util/_decorators.py:
```diff
@@ -65,8 +65,9 @@ def deprecate_kwarg(old_arg_name, new_arg_name, mapping=None, stacklevel=2):
+        Name of preferred argument in function. Use None to raise warning that
+        ``old_arg_name`` keyword is deprecated.
     mapping : dict or callable
         If mapping is present, use it to translate old arguments to
```
could you add a couple of examples to the doc-string here? e.g. doing a rename and removing a kwarg

EricChea replied:
I added a couple examples for removing the keyword.  

I did see a couple good examples of how to raise a warning when a keyword will be deprecated for another:
[keyword to keyword](https://github.com/pandas-dev/pandas/pull/19751/files#diff-e41db70762f6c5b076f9b551651d4132R80)
[keyword map (multiple keyword deprecations)](https://github.com/pandas-dev/pandas/pull/19751/files#diff-e41db70762f6c5b076f9b551651d4132R92)

Do let me know if you think we should still add some more examples for renaming kwargs.

jreback commented on pandas/util/_decorators.py:
```diff
@@ -107,6 +108,16 @@ def _deprecate_kwarg(func):
+
+            if new_arg_name is None and old_arg_value is not None:
+                msg = (
+                    "the '{old_name}' keyword is deprecated and will be removed in a future version"
```
you might need to break this line (too long) for the linter

EricChea replied:
Thanks for catching that.  I need to remember to run ``git diff master -u -- "*.py" | flake8 --diff``

jreback commented on pandas/util/_decorators.py:
```diff
@@ -107,6 +108,16 @@ def _deprecate_kwarg(func):
+                msg = (
+                    "the '{old_name}' keyword is deprecated and will be removed in a future version"
+                    "please takes steps to stop use of '{old_name}'"
+                ).format(old_name=old_arg_name)
```
since we have this new capability for ``deprecate_kwarg`` if you wanted (in a followup PR), to change existing code to use this kwargs rather than the 'manual' way of warning (e.g. before we had this)

EricChea replied:
Sure.  I'll take a look around and open a separate PR.

EricChea replied:
@jreback following up.  I found that the 'manual' deprecation warnings in the repo target an entire function or class (and in some cases class attributes).  I didn't find any for keyword specific deprecations.   I suppose I can create an analogous decorator called ``deprecate_func()`` or ``deprecate_obj`` if that would be of value.

EricChea committed:
Add examples of how to deprecate a keyword, and fix spacing.

jreback committed:
doc

jreback commented:
can you add a whatsnew note in the deprecation section, otherwise lgtm.

TomAugspurger commented on pandas/util/_decorators.py:
```diff
@@ -65,8 +65,9 @@ def deprecate_kwarg(old_arg_name, new_arg_name, mapping=None, stacklevel=2):
         Name of argument in function to deprecate
-    new_arg_name : str
-        Name of preferred argument in function
+    new_arg_name : str | None
```
Usually this is written as `str or None`.

EricChea replied:
Thanks for catching this -- didn't realize this is the case.  Please refer to 4c83659350328784ecdac963f7bb8b6a5acbf577

TomAugspurger commented on pandas/util/_decorators.py:
```diff
@@ -96,6 +100,25 @@ def deprecate_kwarg(old_arg_name, new_arg_name, mapping=None, stacklevel=2):
       warnings.warn(msg, FutureWarning)
     yes!
+
+
```
Can you have this many blank lines in a docstring without flake8 complaining?

EricChea replied:
Yes that is correct.  I didn't get any warnings here.

EricChea committed:
Add to what's new: deprecation of keyword ability, and warning when order is used with functionalize

EricChea committed:
Change '|' to 'or' in docs.

EricChea committed:
Merge branch 'master' into dep_keyword

jorisvandenbossche commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -589,6 +589,7 @@ Other API Changes
 - :func:`Series.to_csv` now accepts a ``compression`` argument that works in the same way as the ``compression`` argument in :func:`DataFrame.to_csv` (:issue:`18958`)
 - Set operations (union, difference...) on :class:`IntervalIndex` with incompatible index types will now raise a ``TypeError`` rather than a ``ValueError`` (:issue:`19329`)
 - :class:`DateOffset` objects render more simply, e.g. ``<DateOffset: days=1>`` instead of ``<DateOffset: kwds={'days': 1}>`` (:issue:`19403`)
+- :func:`util._decorators.deprecate_kwarg` now has the ability to deprecate a keyword entirely. Previously, using ``@deprecate_kwarg`` required that an alternative keyword be provided in favor of the deprecated keyword.
```
It's not needed to add this to the whatsnew docs, since this is private functionality. So the deprecation warning about `order` itself is enough.

EricChea replied:
Makes sense.  Please refer to b5daae00dfae1c714936d49572deb5e985869e00

EricChea committed:
Remove deprecate_kwarg util changes from whats new.

jreback committed:
doc

jreback committed:
Merge branch 'master' into PR_TOOL_MERGE_PR_19751

jreback commented:
thanks @EricChea 

small correction on merge, the note was in the incorrect deprecation section.

EricChea commented:
@jreback thanks for doing that.
